lib/apiservers/engine/backends/container_proxy.go

```diff
@@ -895,3 +1016,167 @@ func portMapFromVicContainer(vc *viccontainer.VicContainer) nat.PortMap {
+	// Pipe for stdin so we can interject and watch the input streams for detach keys.
+	stdinReader, stdinWriter := io.Pipe()
+
+	defer stdinWriter.Close()
+
+	go func() {
+		defer stdinReader.Close()
+		// Copy the stdin from the CLI and write to a pipe.  We need to do this so we can
+		// watch the stdin stream for the detach keys.
+		var err error
+
+		// Write some init bytes into the pipe to force Swagger to make the initial
+		// call to the portlayer, prior to any user input in whatever attach client
+		// he/she is using.
+		stdinWriter.Write([]byte(attachStdinInitString))
+
+		if vc.Config.Tty {
+			_, err = copyEscapable(stdinWriter, clStdin, keys)
+		} else {
+			_, err = io.Copy(stdinWriter, clStdin)
+		}
+
+		if err != nil {
+			log.Errorf("container attach: stdin err: %s", err.Error())
+		}
+	}()
+
+	// Swagger wants an io.reader so give it the reader pipe.  Also, the swagger call
+	// to set the stdin is synchronous so we need to run in a goroutine
+	setStdinParams := interaction.NewContainerSetStdinParamsWithContext(ctx).WithID(vc.ContainerID)
+	setStdinParams = setStdinParams.WithRawStream(stdinReader)
+	_, err := pl.Interaction.ContainerSetStdin(setStdinParams)
+
+	if vc.Config.StdinOnce && !vc.Config.Tty {
+		// Close the stdin connection.  Mimicing Docker's behavior.
+		// FIXME: If we close this stdin connection.  The portlayer does
+		// not really close stdin.  This is diff from current Docker
+		// behavior.  However, we're not sure why Docker even has this
+		// behavior where you connect to stdin on the first time only.
+		// If we really want to add this behavior, we need to add support
+		// in the tether in the portlayer.
+		log.Errorf("Attach stream has stdinOnce set.  VIC does not yet support this.")
+	}
+
+	return err
+}
+
+func copyStdOut(ctx context.Context, pl *client.PortLayer, attemptTimeout time.Duration, vc *viccontainer.VicContainer, clStdout io.Writer) error {
+	name := vc.ContainerID
+	//Calculate how much time to let portlayer attempt
+	plAttemptTimeout := attemptTimeout - attachPLAttemptDiff //assumes personality deadline longer than portlayer's deadline
+	plAttemptDeadline := time.Now().Add(plAttemptTimeout)
+	swaggerDeadline := strfmt.DateTime(plAttemptDeadline)
+	log.Debugf("* stdout portlayer deadline: %s", plAttemptDeadline.Format(time.UnixDate))
+	log.Debugf("* stdout personality deadline: %s", time.Now().Add(attemptTimeout).Format(time.UnixDate))
+
+	log.Debugf("* stdout attach start %s", time.Now().Format(time.UnixDate))
+	getStdoutParams := interaction.NewContainerGetStdoutParamsWithContext(ctx).WithID(name).WithDeadline(&swaggerDeadline)
+	_, err := pl.Interaction.ContainerGetStdout(getStdoutParams, clStdout)
+	log.Debugf("* stdout attach end %s", time.Now().Format(time.UnixDate))
+	if err != nil {
+		if _, ok := err.(*interaction.ContainerGetStdoutNotFound); ok {
+			return NotFoundError(name)
+		}
+		if _, ok := err.(*interaction.ContainerGetStdoutInternalServerError); ok {
```

Comment:
Does it make sense to check this type if it returns InternalServerError that is the same as return after this if?